### PR TITLE
[SW-1298] move fungibleTokens from store to composable

### PR DESCRIPTION
**Pull request by konradodo**

### CODE REVIEW

File: src/composables/fungibleTokens.ts [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+      aci: FungibleTokenFullInterfaceACI,
+      address: selectedToken?.contractId as any,
+    });
+
+    const { decodedResult } = await tokenContract.allowance({
+      from_account: account?.address,
+      for_account: tippingContractAddresses?.value?.tippingV2?.replace('ct_', 'ak_'),
+    });
+
+    const allowanceAmount = decodedResult !== undefined
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:124]:
```suggestion
    const allowanceAmount = (decodedResult !== undefined)
```

File: src/composables/fungibleTokens.ts [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+      ? new BigNumber(decodedResult)
+        .multipliedBy(-1)
+        .plus(toShiftedBigNumber(amount, selectedToken?.decimals!))
+        .toNumber()
+      : toShiftedBigNumber(amount, selectedToken?.decimals!).toFixed();
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:129]: one will return number and the other one string - please adjust the types

File: src/composables/fungibleTokens.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+    const getContractFunction = (tokenContract.methods as any)[
+      decodedResult !== undefined ? 'change_allowance' : 'create_allowance'
+    ];
+
+    return getContractFunction(
+      tippingContractAddresses.value?.tippingV2?.replace('ct_', 'ak_'),
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:136]: 'ct_' and 'ak_' could be retrieved from `Encoding`

File: src/composables/fungibleTokens.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    } catch (error) {
+      return {};
+    }
+  }
+
+  async function transfer(
+    address: string,
+    toAccount: any,
+    amount: any,
+    option: any,
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:198]: Are you able to provide some types here?

File: src/composables/fungibleTokens.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    address: string,
+    posAddress: any,
+    invoiceId: any,
+    amount: any,
+    option: any,
+  ) {
+    const aeSdk = await getAeSdk();
+    const tokenContract = await aeSdk.initializeContract({
+      aci: ZeitTokenACI,
+      address: address as any,
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:218]: IDEM

File: src/composables/fungibleTokens.ts [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  }
+
+  watch(activeNetwork, async (network, oldNetwork) => {
+    const newMiddlewareUrl = network.protocols[PROTOCOL_AETERNITY].middlewareUrl;
+    const oldMiddlewareUrl = oldNetwork?.protocols?.[PROTOCOL_AETERNITY]?.middlewareUrl;
+    if (newMiddlewareUrl === oldMiddlewareUrl) {
+      return;
+    }
+    await loadAvailableTokens();
+    await loadTokenBalances();
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:232]:
```suggestion
    if (newMiddlewareUrl !== oldMiddlewareUrl) {
      await loadAvailableTokens();
      loadTokenBalances();
    }
```

File: src/composables/fungibleTokens.ts [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+  }, { immediate: true });
+
+  watch(aeNextAccountIdx, async (val, oldVal) => {
+    if (val !== oldVal) {
+      await loadTokenBalances();
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:237]: I think that `await` is not needed here
```suggestion
     loadTokenBalances();
```

File: src/composables/fungibleTokens.ts [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+      return +toShiftedBigNumber(
+        contractCallData.amount || 0,
+        -availableTokens.value[contractCallData.token!].decimals,
+      );
+    }
+    const isReceived = direction === TX_DIRECTION.received;
+
+    const rawAmount = transaction.tx?.amount
+            || (transaction.tx?.tx?.tx as any)?.amount
+            || transaction.tx?.nameFee || 0;
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:264]:
```suggestion
    const rawAmount = (transaction.tx?.amount)
      || (transaction.tx?.tx?.tx as any)?.amount
      || transaction.tx?.nameFee || 0;
```

File: src/composables/pollAccountInfo.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  function startPolling() {
+    pollDefaultNames = setInterval(() => store.dispatch('names/setDefaults'), 10000);
+    pollAvailableTokens = setInterval(() => loadAvailableTokens(), 60000);
+    pollTokenBalances = setInterval(() => loadTokenBalances(), 10000);
+  }
+
+  function handleVisibilityChange(event: Event) {
+    const isHidden = (event.target as any).hidden;
```
Comment (pstachel) [inline src/composables/pollAccountInfo.ts:28]: missing type

File: src/composables/pollAccountInfo.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    clearInterval(pollTokenBalances);
+  }
+
+  function startPolling() {
+    pollDefaultNames = setInterval(() => store.dispatch('names/setDefaults'), 10000);
+    pollAvailableTokens = setInterval(() => loadAvailableTokens(), 60000);
+    pollTokenBalances = setInterval(() => loadTokenBalances(), 10000);
+  }
+
+  function handleVisibilityChange(event: Event) {
```
Comment (pstachel) [inline src/composables/pollAccountInfo.ts:27]: I think you could get rid of this metod in favour of `createPollingBasedOnMountedComponents`

File: src/popup/components/SwapRoute.vue [hunk 11]
```diff
@@ -30,78 +30,98 @@
...
+      if (tokens[0].isAe && waeContract && !waeContract?.includes(tokens[1].contractId as any)) {
         tokens.unshift({
           ...tokens[0],
           isAe: true,
         });
         tokens[1].isAe = false;
       }
       if (tokens[tokens.length - 1].isAe && waeContract
-        && !waeContract?.includes(tokens[tokens.length - 2].contractId)) {
+        && !waeContract?.includes(tokens[tokens.length - 2].contractId as any)) {
```
Comment (pstachel) [inline src/popup/components/SwapRoute.vue:97]: If you are dropping one condition all of them should be dropped

File: src/popup/components/SwapRoute.vue [hunk 12]
```diff
@@ -30,78 +30,98 @@
...
+      if (idx === tokensList.value.length - 1) {
         return false;
       }
-      const { nodeNetworkId } = useAeSdk({ store: this.$store });
-      const contracts = DEX_CONTRACTS[nodeNetworkId.value];
+      const contracts = DEX_CONTRACTS[nodeNetworkId.value!];
       return (
-        contracts?.wae?.includes(this.tokens[idx].contractId)
-        && this.tokens[idx + 1].isAe
+        contracts?.wae?.includes(tokensList.value[idx].contractId as any)
```
Comment (pstachel) [inline src/popup/components/SwapRoute.vue:110]: isn't the type for contractId `Encoding.ContractAddress`?

File: src/popup/components/TransactionList.vue [hunk 13]
```diff
@@ -50,10 +50,9 @@ import {
   pipe,
   sortTransactionsByDate,
 } from '@/utils';
-import { useGetter } from '@/composables/vuex';
 import {
   useAccounts,
-  useAeSdk,
+  useAeSdk, useFungibleTokens,
```
Comment (pstachel) [inline src/popup/components/TransactionList.vue:55]: please drop this line

File: src/popup/pages/MultisigProposalDetails.vue [hunk 14]
```diff
@@ -303,9 +303,8 @@ import {
   useMultisigAccounts,
   usePendingMultisigTransaction,
   useMultisigTransactions,
-  useModals,
+  useModals, useFungibleTokens,
```
Comment (pstachel) [inline src/popup/pages/MultisigProposalDetails.vue:306]: please drop this line

File: src/popup/pages/Retip.vue [hunk 15]
```diff
@@ -92,7 +92,7 @@ import {
   useAccounts,
   useTippingContracts,
   useAeSdk,
-  useTransactionList,
+  useTransactionList, useFungibleTokens,
```
Comment (pstachel) [inline src/popup/pages/Retip.vue:95]: Please drop this line

File: src/protocols/aeternity/components/TransferReview.vue [hunk 16]
```diff
@@ -94,7 +94,7 @@ import {
 } from '@/utils';
 import {
   useAccounts,
-  useDeepLinkApi,
+  useDeepLinkApi, useFungibleTokens,
```
Comment (pstachel) [inline src/protocols/aeternity/components/TransferReview.vue:97]:
```suggestion
  useDeepLinkApi,
   useFungibleTokens,
```

File: src/protocols/aeternity/components/TransferSendForm.vue [hunk 17]
```diff
@@ -176,7 +176,7 @@ import {
 import {
   useAccounts,
   useBalances,
-  useCurrencies,
+  useCurrencies, useFungibleTokens,
```
Comment (pstachel) [inline src/protocols/aeternity/components/TransferSendForm.vue:179]:
```suggestion
  useCurrencies,
  useFungibleTokens,
```

File: src/protocols/aeternity/views/TransactionDetails.vue [hunk 18]
```diff
@@ -63,7 +63,7 @@ import { useRoute, useRouter } from 'vue-router';
 import { Encoded, Tag } from '@aeternity/aepp-sdk';
 import type { ITransaction, TxFunctionRaw } from '@/types';
 import {
-  useAccounts,
+  useAccounts, useFungibleTokens,
```
Comment (pstachel) [inline src/protocols/aeternity/views/TransactionDetails.vue:66]:
```suggestion
  useAccounts,
  useFungibleTokens,
```

File: src/store/plugins/pollAccountInfo.js
Comment (pstachel) [inline src/store/plugins/pollAccountInfo.js:1]: Is this file still needed?

File: tests/unit/getters/get-tx-amount-total.spec.js [hunk 19]
```diff
@@ -1,85 +1,89 @@
-import BigNumber from 'bignumber.js';
-import getters from '../../../src/store/getters';
-import { TX_DIRECTION } from '../../../src/constants';
-import { STUB_TOKEN_CONTRACT_ADDRESS, STUB_TRANSACTIONS } from '../../../src/constants/stubs';
-import { AE_COIN_PRECISION } from '../../../src/protocols/aeternity/config';
+// import BigNumber from 'bignumber.js';
```
Comment (pstachel) [inline tests/unit/getters/get-tx-amount-total.spec.js:1]: I think you should uncomment that

**Review by pstachel — CHANGES_REQUESTED** [review]

File: src/composables/fungibleTokens.ts [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+import ZeitTokenACI from '@/lib/contracts/FungibleTokenFullACI.json';
+import { AE_SYMBOL } from '@/protocols/aeternity/config';
+import { useAccounts } from './accounts';
+import { useAeSdk } from './aeSdk';
+import { useMiddleware } from './middleware';
+import { useTippingContracts } from './tippingContracts';
+import { useNetworks } from './networks';
+
+const availableTokens = ref<ITokenList>({});
+const tokensBalance = ref<Record<string, IToken[]>>({});
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:26]: As this is a record (list) I suggest to use plural of "balance" (balances).

File: src/composables/fungibleTokens.ts [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+      };
+    }, {});
+  }
+
+  async function loadTokenBalances() {
+    if (!isLoggedIn.value) {
+      return;
+    }
+    const addresses = (accounts.value)
+      .filter((account) => account.protocol === PROTOCOL_AETERNITY)
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:71]: Accounts composable returns this list as `aeAccounts`.

File: src/composables/fungibleTokens.ts [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+            const convertedBalance = Number(balance.toFixed(2));
+
+            return {
+              ...availableToken,
+              value: token.contractId,
+              text: `${convertedBalance} ${availableToken.symbol}`,
+              contractId: token.contractId,
+              balance,
+              convertedBalance,
+            } as IToken;
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:98]: If you use `as Something` the TS engine won't fully check the data. I suggest to first create a typed constant and then return it:
```
const token: IToken = { ... };
return token;
```

File: src/composables/fungibleTokens.ts [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+    const tokenContract = await aeSdk.initializeContract({
+      aci: ZeitTokenACI,
+      address: address as any,
+    });
+    return tokenContract.burn_trigger_pos(
+      amount.toFixed(), posAddress, invoiceId, option,
+    );
+  }
+
+  watch(activeNetwork, async (network, oldNetwork) => {
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:225]: Previous watcher was initiated at the app runtime and was present all the time. This one will be removed each time components using the composable are unmounted. So maybe it is better to use `onNetworkChange` present in networks composable?

File: src/composables/fungibleTokens.ts [hunk 24]
```diff
@@ -0,0 +1,288 @@
...
+    await loadTokenBalances();
+  }, { immediate: true });
+
+  watch(aeNextAccountIdx, async (val, oldVal) => {
+    if (val !== oldVal) {
+      await loadTokenBalances();
+    }
+  });
+
+  function getTxSymbol(transaction?: ITransaction) {
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:241]: Methods should be placed before watchers.

File: src/composables/fungibleTokens.ts [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+    if (transaction?.pendingTokenTx) {
+      return availableTokens.value[transaction.tx.contractId]?.symbol;
+    }
+    const contractCallData = transaction?.tx && categorizeContractCallTxObject(transaction);
+    return availableTokens.value[contractCallData?.token!]?.symbol || AE_SYMBOL;
+  }
+
+  function getTxAmountTotal(
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:249]: Methods should be placed before watchers.

File: src/composables/fungibleTokens.ts (hunk 8 above)
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:264]: Indentation issues.

File: src/store/plugins/pollAccountInfo.js [hunk 26]
```diff
@@ -11,8 +18,8 @@ export default async (store) => {
 
   function startPolling() {
     pollDefaultNames = setInterval(() => store.dispatch('names/setDefaults'), 10000);
-    pollAvailableTokens = setInterval(() => store.dispatch('fungibleTokens/loadAvailableTokens'), 60000);
-    pollTokenBalances = setInterval(() => store.dispatch('fungibleTokens/loadTokenBalances'), 10000);
+    pollAvailableTokens = setInterval(() => loadAvailableTokens(), 60000);
+    pollTokenBalances = setInterval(() => loadTokenBalances(), 10000);
```
Comment (peronczyk) [inline src/store/plugins/pollAccountInfo.js:22]: I think instead of constant polling we should use composable helper: `createPollingBasedOnMountedComponents` within the `fungibleTokens` composable so the polling will only happen when any of the App components would actually need this.

File: tests/unit/getters/get-tx-amount-total.spec.js
Comment (peronczyk) [inline tests/unit/getters/get-tx-amount-total.spec.js:1]: I guess the tests needs to be aligned with the introduced changes.

File: src/composables/fungibleTokens.ts [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+import { useAeSdk } from './aeSdk';
+import { useMiddleware } from './middleware';
+import { useTippingContracts } from './tippingContracts';
+import { useNetworks } from './networks';
+
+const availableTokens = ref<ITokenList>({});
+const tokensBalance = ref<Record<string, IToken[]>>({});
+
+export function useFungibleTokens() {
+  const store = useStore();
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:29]: `useStore` should not be used out of the component's scope. I suggest to pass the store as the argument the same way as othe composables.

File: src/composables/fungibleTokens.ts [hunk 28]
```diff
@@ -0,0 +1,288 @@
+import { ref, watch } from 'vue';
+import { useStore } from 'vuex';
+import camelCaseKeysDeep from 'camelcase-keys-deep';
+import BigNumber from 'bignumber.js';
+import { Encoded } from '@aeternity/aepp-sdk';
+import { fetchAllPages, handleUnknownError, toShiftedBigNumber } from '@/utils';
+import {
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:7]: Please move types before any local import and use `import type`.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/composables/pollAccountInfo.ts (hunk 10 above)
Comment (peronczyk) [reply src/composables/pollAccountInfo.ts:27]: Totally

File: src/popup/components/SwapRoute.vue [hunk 29]
```diff
@@ -30,78 +30,98 @@
...
       )
-      || (
-        contracts?.wae?.includes(this.tokens[idx + 1].contractId)
-        && this.tokens[idx].isAe
-      );
-    },
+        || (
+          contracts?.wae?.includes(tokensList.value[idx + 1].contractId as any)
+          && tokensList.value[idx].isAe
+        );
```
Comment (peronczyk) [inline src/popup/components/SwapRoute.vue:116]: Wrong indentation.

File: src/popup/components/SwapRoute.vue [hunk 30]
```diff
@@ -30,78 +30,98 @@
...
+    const { nodeNetworkId } = useAeSdk({ store });
+    const { availableTokens } = useFungibleTokens();
+
+    function getTxFunction(
+      functionName: TxFunctionRaw | TxFunctionParsed | TxFunction,
+    ): TxFunctionParsed {
+      return camelCase(functionName) as TxFunctionParsed;
+    }
+
+    const tokensList = computed(() => {
```
Comment (peronczyk) [inline src/popup/components/SwapRoute.vue:66]:
```suggestion
    const tokenList = computed(() => {
```

File: src/popup/components/TransactionLabel.vue [hunk 31]
```diff
@@ -78,12 +78,8 @@ import { Tag } from '@aeternity/aepp-sdk';
...
 import { useStore } from 'vuex';
 
-import { useAccounts, useTransactionTx } from '@/composables';
-import { useState } from '@/composables/vuex';
-import {
-  ITokenList,
-  ITransaction,
-} from '@/types';
+import { useAccounts, useFungibleTokens, useTransactionTx } from '@/composables';
+import { ITransaction } from '@/types';
```
Comment (peronczyk) [inline src/popup/components/TransactionLabel.vue:82]:
```suggestion
import type { ITransaction } from '@/types';
```

File: src/composables/pollAccountInfo.ts [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+    loadAvailableTokens,
+  } = useFungibleTokens({ store });
+
+  const defaultNamesWatcher = createPollingBasedOnMountedComponents(10000);
+  const availableTokensWatcher = createPollingBasedOnMountedComponents(60000);
+  const tokenBalancesWatcher = createPollingBasedOnMountedComponents(10000);
+
+  function startPolling() {
+    defaultNamesWatcher(() => store.dispatch('names/setDefaults'));
+    availableTokensWatcher(() => loadAvailableTokens());
```
Comment (CedrikNikita) [inline src/composables/pollAccountInfo.ts:19]: I think within this PR we can consider not loading available tokens on initialization.
- all the tokens are only needed in receive modal token list
- currently token balances endpoint contains all the needed info to show them, without loading all available tokens
- we would need to add new method load all the needed info for a particular token in several places

Comment (konradodo) [reply src/composables/pollAccountInfo.ts:19]: https://aeternity-blockchain.atlassian.net/jira/software/c/projects/SW/boards/2?selectedIssue=SW-1330

Comment (konradodo) [reply src/composables/pollAccountInfo.ts:19]: i prepared new task for it

File: src/composables/fungibleTokens.ts [hunk 33]
```diff
@@ -0,0 +1,300 @@
...
+const { onNetworkChange } = createNetworkWatcher();
+
+export function useFungibleTokens({ store }: IDefaultComposableOptions) {
+  const { getAeSdk } = useAeSdk({ store });
+  const { fetchFromMiddleware } = useMiddleware();
+  const { tippingContractAddresses } = useTippingContracts({ store });
+  const {
+    isLoggedIn,
+    aeAccounts,
+    aeNextAccountIdx,
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:36]: This will be removed within this PR: https://github.com/aeternity/superhero-wallet/pull/2334
I strongly suggest to rebase your branch on top of my as it is already reviewed+tested and will be merged right after the Ionic release.

File: src/composables/pollAccountInfo.ts (hunk 32 above)
Comment (peronczyk) [reply src/composables/pollAccountInfo.ts:19]: I don't think creating separate task for this suggestion is a good thing. Actually @pstachel , @CedrikNikita and me made the same comment and I'm almost 100% sure we all can agree that this composable file is a wrong implementation of `createPollingBasedOnMountedComponents`.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/composables/fungibleTokens.ts [hunk 34]
```diff
@@ -0,0 +1,310 @@
...
+      loadTokenBalances();
+    }
+  });
+
+  (() => {
+    if (!isInitialized) {
+      availableTokensWatcher(() => loadAvailableTokens());
+      tokenBalancesWatcher(() => loadTokenBalances());
+    }
+  })();
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:295]:
```suggestion
  availableTokensWatcher(() => loadAvailableTokens());
  tokenBalancesWatcher(() => loadTokenBalances());
```
Polling watcher should be fired every time the composable is used. This allows it to count how many components are using this composable.

File: src/composables/fungibleTokens.ts [hunk 35]
```diff
@@ -0,0 +1,310 @@
...
+        totalSupply,
+        balance,
+        balances,
+      };
+    } catch (error) {
+      return {};
+    }
+  }
+
+  async function transfer(
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:198]: Previously this method was used as `dispatch('fungibleTokens/transfer', ...)` so it was clear what kind of transfer we are making. Now the method `transfer` is not clear enough in my opinion. I suggest to rename it to `transferToken`.

File: src/composables/fungibleTokens.ts [hunk 36]
```diff
@@ -0,0 +1,310 @@
...
+    address: Encoded.ContractAddress,
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:199]: To avoid mistaking this with user address I suggest to rename this argument to `tokenContractId`.

File: src/composables/tokensList.ts [hunk 37]
```diff
@@ -14,6 +14,8 @@ import { AE_CONTRACT_ID } from '@/protocols/aeternity/config';
 import { ProtocolAdapterFactory } from '@/lib/ProtocolAdapterFactory';
 import { PROTOCOL_AETERNITY } from '@/constants';
 import { useCurrencies } from '@/composables/currencies';
+import { useFungibleTokens } from '@/composables/fungibleTokens';
+import { useAccounts } from '@/composables/accounts';
```
Comment (peronczyk) [inline src/composables/tokensList.ts:18]: As those files are located in the same directory I suggest to use `./`.

File: src/composables/transactionTokens.ts [hunk 38]
```diff
@@ -20,6 +19,7 @@ import {
 } from '@/protocols/aeternity/helpers';
 import { BTC_SYMBOL } from '@/protocols/bitcoin/config';
 import { getTxAmountTotal as getBitcoinTxAmountTotal } from '@/protocols/bitcoin/helpers';
+import { useFungibleTokens } from '@/composables/fungibleTokens';
```
Comment (peronczyk) [inline src/composables/transactionTokens.ts:22]:
```suggestion
import { useFungibleTokens } from './fungibleTokens';
```

File: src/popup/App.vue [hunk 39]
```diff
@@ -107,6 +108,7 @@ export default defineComponent({
     const { addWalletNotification } = useNotifications({ store });
     const { loadCoinsData } = useCurrencies({ withoutPolling: true });
     const { initViewport } = useViewport();
+    useFungibleTokens({ store });
```
Comment (peronczyk) [inline src/popup/App.vue:111]: We need to choose one of two options:
1. Fetch fungible tokens only if they are needed
2. Poll them all the time

With this line it is this weird mix of those two. The composable is using polling based on mounted components, which means we want to poll only when needed. But here we are loading the composable globally so the polling will be happening all the time as there is always at least one component that is using the composable.

My suggestion: remove this line.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/popup/components/TransactionList.vue [hunk 40]
```diff
@@ -140,7 +140,7 @@ export default defineComponent({
         (transaction) => (
           !searchPhrase.value
           || includesCaseInsensitive(
-            getTxSymbol.value(transaction),
+            getTxSymbol(transaction as ITransaction),
```
Comment (peronczyk) [inline src/popup/components/TransactionList.vue:143]: Instead of using `as` here I suggest to allow to pass `ICommonTransaction` as the argument of `getTxSymbol`.

File: src/store/getters.js [hunk 41]
```diff
@@ -1,15 +1,52 @@
...
+              idx,
+            )
+          : {}
+        ),
+      }))
+      .map(({ ...account }) => ({
+        ...account,
+        name: getters['names/getDefault'](account.address),
+      }));
+  },
```
Comment (peronczyk) [inline src/store/getters.js:49]: This was removed within the `accounts` migration from Vuex to composable. Please delete this block.

File: src/store/getters.js [hunk 42]
```diff
@@ -1,15 +1,52 @@
...
   getTxSymbol: ({ fungibleTokens: { availableTokens } }) => (transaction) => {
```
Comment (peronczyk) [inline src/store/getters.js:50]: This can be removed because you created same method within the `fungibleTokens` composable.

File: src/store/getters.js
Comment (peronczyk) [inline src/store/getters.js:1]: Probably this whole file can be removed now.

**Review by peronczyk — CHANGES_REQUESTED** [review]

File: src/composables/fungibleTokens.ts [hunk 43]
```diff
@@ -0,0 +1,307 @@
...
+import { useTippingContracts } from './tippingContracts';
+import { createNetworkWatcher } from './networks';
+import { createPollingBasedOnMountedComponents } from './composablesHelpers';
+
+const availableTokens = ref<ITokenList>({});
+const tokenBalances = ref<Record<string, IToken[]>>({});
+
+const { onNetworkChange } = createNetworkWatcher();
+const availableTokensWatcher = createPollingBasedOnMountedComponents(60000);
+const tokenBalancesWatcher = createPollingBasedOnMountedComponents(10000);
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:30]: Those functions are not watchers for either availableTokens nor tokenBalances - they initialises polling

File: src/composables/fungibleTokens.ts [hunk 44]
```diff
@@ -0,0 +1,307 @@
...
+  watch(aeAccounts, (val, oldVal) => {
+    if (val !== oldVal) {
+      loadTokenBalances();
+    }
+  });
+
+  (() => {
+    availableTokensWatcher(() => loadAvailableTokens());
+    tokenBalancesWatcher(() => loadTokenBalances());
+  })();
```
Comment (pstachel) [inline src/composables/fungibleTokens.ts:292]:
```suggestion
  availableTokensWatcher(() => loadAvailableTokens());
  tokenBalancesWatcher(() => loadTokenBalances());
```

File: src/popup/components/Modals/ConfirmTransactionSign.vue [hunk 45]
```diff
@@ -276,13 +274,13 @@ export default defineComponent({
...
     const totalAmount = computed(
-      () => getTxAmountTotal.value(transactionWrapped.value, direction.value),
+      () => getTxAmountTotal(transactionWrapped.value as ITransaction, direction.value),
     );
 
     const singleToken = computed((): ITokenResolved => ({
       isReceived: direction.value === TX_DIRECTION.received,
       amount: totalAmount.value,
-      symbol: getTxSymbol.value(popupProps.value?.tx),
+      symbol: getTxSymbol(popupProps.value?.tx as any),
```
Comment (pstachel) [inline src/popup/components/Modals/ConfirmTransactionSign.vue:283]:
```suggestion
      symbol: getTxSymbol(popupProps.value?.tx as ITx),
```

File: src/popup/components/SwapRoute.vue [hunk 46]
```diff
@@ -30,78 +30,104 @@
...
           tokens[1],
         ];
       }
-      const { nodeNetworkId } = useAeSdk({ store: this.$store });
-      const waeContract = DEX_CONTRACTS[nodeNetworkId.value]?.wae;
-      if (tokens[0].isAe && waeContract && !waeContract?.includes(tokens[1].contractId)) {
+
+      const waeContract = DEX_CONTRACTS[nodeNetworkId.value!]?.wae;
+
+      if (tokens[0].isAe && waeContract && !waeContract?.includes(tokens[1].contractId as any)) {
```
Comment (pstachel) [inline src/popup/components/SwapRoute.vue:93]: I think we are able to type contractId

File: src/popup/components/SwapRoute.vue [hunk 47]
```diff
@@ -30,78 +30,104 @@
...
+
+    const tokenList = computed(() => {
+      const txFunction = getTxFunction(props.transaction.tx.function!);
+
+      if (!isTxFunctionDexSwap(txFunction)) {
         return [];
       }
-      const resolver = getTransactionTokenInfoResolver(camelCase(this.transaction.tx.function));
+      const resolver = getTransactionTokenInfoResolver(txFunction);
       if (!resolver) return [];
```
Comment (pstachel) [inline src/popup/components/SwapRoute.vue:77]:
```suggestion
      if (!resolver) {
        return [];
      }
```

File: src/popup/components/SwapRoute.vue [hunk 48]
```diff
@@ -30,78 +30,104 @@
...
-      return (
-        contracts?.wae?.includes(this.tokens[idx].contractId)
-        && this.tokens[idx + 1].isAe
-      )
-      || (
-        contracts?.wae?.includes(this.tokens[idx + 1].contractId)
-        && this.tokens[idx].isAe
+      const contracts = DEX_CONTRACTS[nodeNetworkId.value!];
+
+      return (contracts?.wae?.includes(tokenList.value[idx].contractId as Encoded.ContractAddress)
```
Comment (pstachel) [inline src/popup/components/SwapRoute.vue:117]: I see that contractId is typed here multiple times. Couldn't you type it inside of the tokenList?

File: src/protocols/aeternity/components/TransferReview.vue [hunk 49]
```diff
@@ -148,6 +149,11 @@ export default defineComponent({
       updateMultisigAccounts,
     } = useMultisigAccounts({ store });
     const { getTippingContracts } = useTippingContracts({ store });
+    const {
+      createOrChangeAllowance,
+      burnTriggerPoS,
+      transferToken: fungibleTokenTransfer,
```
Comment (peronczyk) [inline src/protocols/aeternity/components/TransferReview.vue:155]: I think there is no need now to rename the method as `transferToken` is fully understandable and does not exist in the scope.

File: src/popup/components/SwapRoute.vue (hunk 48 above)
Comment (konradodo) [reply src/popup/components/SwapRoute.vue:117]: IToken have this type`
  contractId: Encoded.ContractAddress
  | typeof AE_CONTRACT_ID
  | typeof BTC_CONTRACT_ID
`

**Review by pstachel — APPROVED** [review]

File: tests/unit/getters/get-tx-amount-total.spec.js [hunk 50]
```diff
@@ -52,31 +53,28 @@ const tests = [{
...
+  const { availableTokens, getTxAmountTotal } = useFungibleTokens({ store });
+
+  availableTokens.value[STUB_TOKEN_CONTRACT_ADDRESS] = {
+    decimals: TEST_TOKEN_DECIMALS,
+  };
+
   tests.forEach((test) => it(
     `should return correct result for each type of transaction: for \
-${test.transaction.tx.type}/${test.transaction.tx.function}`,
+          ${test.transaction.tx.type}/${test.transaction.tx.function}`,
```
Comment (CedrikNikita) [inline tests/unit/getters/get-tx-amount-total.spec.js:66]: Name of the test before this changes:
`should return correct result for each type of transaction: for PayingForTx/undefined`
name of the test after this changes:
`should return correct result for each type of transaction: for           PayingForTx/undefined`

File: src/composables/fungibleTokens.ts [hunk 51]
```diff
@@ -0,0 +1,317 @@
...
+  const { getAeSdk } = useAeSdk({ store });
+  const { fetchFromMiddleware } = useMiddleware();
+  const { tippingContractAddresses } = useTippingContracts({ store });
+  const {
+    isLoggedIn,
+    aeAccounts,
+    getLastActiveProtocolAccount,
+  } = useAccounts();
+
+  function getTokenBalance(address?: string) {
```
Comment (CedrikNikita) [inline src/composables/fungibleTokens.ts:62]: Why the method is called in singular (not in plural), but returning all the balances for the given address?

File: src/popup/components/AccountItem.vue [hunk 52]
```diff
@@ -34,7 +34,6 @@
       v-else
       class="address"
       :address="address"
-      :protocol="protocol"
```
[the remark this answers is not in the document]
Comment (peronczyk) [reply src/popup/components/AccountItem.vue:37]: That's a mistake I think. I'll bring this back.

File: src/composables/fungibleTokens.ts (hunk 51 above)
Comment (peronczyk) [reply src/composables/fungibleTokens.ts:62]: Totally makes sense. I think I've made the same comment to Konrad but he didn't include any fix. I changed the method name to `getAccountTokenBalances`.

File: tests/unit/getters/get-tx-amount-total.spec.js (hunk 50 above)
Comment (peronczyk) [reply tests/unit/getters/get-tx-amount-total.spec.js:66]: OK< I removed the white space.

File: src/popup/components/AccountSelector.vue [hunk 53]
```diff
@@ -41,7 +41,6 @@
       <AddressTruncated
         v-if="!avatarOnly"
         show-explorer-link
-        show-protocol-icon
```
Comment (CedrikNikita) [inline src/popup/components/AccountSelector.vue:44]: What about this prop? Was it also removed by mistake?

Comment (peronczyk) [reply src/popup/components/AccountSelector.vue:44]: To be honest I'm not sure. But as the task's scope should not cover this change I'm bringing this line back.

**Review by CedrikNikita — APPROVED** [review]
Code changes look reasonable. Should be align with https://github.com/superhero-com/superhero-wallet/pull/2409 or vise verse, because the `useAeSdk` is no longer needed to use store on initialization.
